radix-engine/src/system/transaction/multithread_txn_processor.rs

```diff
@@ -0,0 +1,138 @@
+
+        // Setup
+        let intents = executable.intents();
+        for (thread_id, intent) in intents.iter().enumerate() {
+            api.kernel_switch_stack(thread_id)?;
+
+            let mut system_service = SystemService::new(api);
+            let virtual_resources = intent
+                .auth_zone_init
+                .simulate_every_proof_under_resources
+                .clone();
+            let virtual_non_fungibles =
+                intent.auth_zone_init.initial_non_fungible_id_proofs.clone();
+            let auth_zone = AuthModule::create_auth_zone(
+                &mut system_service,
+                None,
+                virtual_resources,
+                virtual_non_fungibles,
+            )?;
+
+            api.kernel_set_call_frame_data(Actor::Function(FunctionActor {
+                blueprint_id: BlueprintId::new(
+                    &TRANSACTION_PROCESSOR_PACKAGE,
+                    TRANSACTION_PROCESSOR_BLUEPRINT,
+                ),
+                ident: TRANSACTION_PROCESSOR_RUN_IDENT.to_string(),
+                auth_zone,
+            }))?;
+
+            let mut system_service = SystemService::new(api);
+            let txn_processor = TxnProcessorThread::<InstructionV2>::init(
+                intent.encoded_instructions.clone(),
+                global_address_reservations.clone(),
+                intent.blobs.clone(),
+                MAX_TOTAL_BLOB_SIZE_PER_INVOCATION,
+                &mut system_service,
+            )?;
+
+            txn_processors.push((txn_processor, intent.children_intent_indices.clone()));
+        }
+        Ok(Self {
+            threads: txn_processors,
+        })
+    }
+
+    pub fn execute<Y: SystemBasedKernelApi>(&mut self, api: &mut Y) -> Result<(), RuntimeError> {
+        let mut cur_thread = 0;
+        let mut parent_stack = vec![];
+
+        loop {
+            api.kernel_switch_stack(cur_thread)?;
+            let (txn_thread, children_mapping) = self.threads.get_mut(cur_thread).unwrap();
+
+            let mut system_service = SystemService::new(api);
+            match txn_thread.resume(&mut system_service)? {
+                ResumeResult::YieldToChild(child) => {
+                    let child = *children_mapping.get(child).unwrap();
+                    parent_stack.push(cur_thread);
+                    cur_thread = child;
+                }
+                ResumeResult::YieldToParent => {
+                    cur_thread = parent_stack.pop().unwrap();
+                }
+                ResumeResult::Done => {
+                    if let Some(parent) = parent_stack.pop() {
+                        cur_thread = parent;
+                    } else {
+                        break;
+                    }
+                }
+            }
+        }
+
+        assert!(parent_stack.is_empty());
+
+        Ok(())
+    }
+
+    pub fn cleanup<Y: SystemBasedKernelApi>(self, api: &mut Y) -> Result<(), RuntimeError> {
```

Comment:
I think I'd rather run `cleanup` greedily when each subintent terminates. I think it gives a more intuitive / timely orphaned node error, if a user is thinking of the intents like stack frames.